### ISSUE TITLE
demo not working  when using geo coordinates based test field. 

### ISSUE THREAD

**Issue by NK1807** [open-navigation/opennav_coverage#40]
Hello @SteveMacenski and team ,
I have been testing demo_row_coverage example and working well with cartesian test filed. But when running the demo with irregular _test_field.xml ( https://github.com/open-navigation/opennav_coverage/blob/main/opennav_coverage/test/irregular_test_field.xml ), it's not as expected and gives out an error as shown below.

[component_container_isolated-4] [INFO] [1707987164.529318727] [row_coverage_server]: Generating coverage path in EPSG:4258 frame for field of 10 rows (9 swaths).
[component_container_isolated-4] [INFO] [1707987164.621530841] [controller_server]: Received a goal, begin computing control effort.
[component_container_isolated-4] [ERROR] [1707987164.621992034] [controller_server]: Resulting plan has 0 poses in it.
[component_container_isolated-4] [WARN] [1707987164.622049720] [controller_server]: [follow_path] [ActionServer] Aborting handle.
[component_container_isolated-4] [ERROR] [1707987164.650015999] [bt_navigator]: Goal failed
[component_container_isolated-4] [WARN] [1707987164.650070816] [bt_navigator]: [navigate_complete_coverage] [ActionServer] Aborting handle.
[demo_row_coverage-7] Task with failed with status code: 6
[demo_row_coverage-7] Goal failed!

I have already replaced the field file in the demo_row_coverage.py with irregular_test_field. and also set coordinates_in_cartesian_frame to  false in demo_params.yaml

Am I missing any other parameters to be set or should i do some workaround for that?

working environment:
Ubuntu 22.04
ROS Iron

**Comment by SteveMacenski** [comment]
Did you add your gps localization, adjust the starting pose for the robot, and all the other necessary changes to make the demo work for another arbitrary field?

does the actual action server for coverage work? I expect so, so need to adjust the things around it.

**Comment by NK1807** [comment]
Yes, I tried adding gps location to the world. I also tested changing the intial pose of the robot near to the field. It shows Goal succeeded in no time(expected behavior). I think I need to do some work around. Thanks for the clarification.

Yes, no significant issues are found with server so all seems good with that.
I'll keep you posted.

**Comment by SteveMacenski** [comment]
Note that if you provide a GPS field, the path published I believe will also be with respect to GPS coordinates, which this simple demo giving the GPS coordinates to the controller won't necessarily work with. Check the data that is publishing out. If working with GPS the expectation is that you're doing other things than the most simplistic demo that is shown in the demo package.

**Comment by NK1807** [comment]
HI @SteveMacenski ,
The path is being publlised in cartesian CS (transformed to UTM) after providing the GPS field and adding GPS localisation but the values are large. I also changed the robot intial pose w.r.t the path published. I tried various approaches like changing controller but the no positive outcome. I'm confused with this frame "EPSG:4258" as in the fields2cover package this is harcoded as it is standard but is there any relation between this frame and the gps sensor frame? or it's just a frame to represent GPS field.

@SteveMacenski , if possible and if available, Could you please provide me some documentation/tutorial to execute GPS based demo? it would be of a great help.

also, if there is any approach that you want to suggest, I'm pretty much open to do that. Thanks in advance.

**Comment by SteveMacenski** [comment]
Again, the demo is built for a cartesian field that is specifically chosen to have the robot model start at the beginning of the path that the coverage generates. There's no reason you should expect that demo works for your arbitrary other field, regardless if its in GPS or not. Please read over the demo code and you'll understand, its just a _demo_.

I can't tell you if the coordinates are realistic or not, its based on the coordinate system and inputs you provided it.. Coordinate system in = coordinate system out.

**Comment by mattiadutto** [comment]
@NK1807 Did you find a solution for using the GPS coordinate with the plugin? Thanks in advice

**Comment by SteveMacenski** [comment]
The demo takes in the Nav Path part of the message to execute. The coordinate system in = coordinate system out. Thus, if you want to use Nav2, your coordinate system in needs to be the cartesian coordinate system of navigation -- and if you're not starting at the edge of the field where the coverage path starts, navigating to that pose first before executing it. 

If you want to use full GPS, you need to provide your own capability to navigate using that information -- or later convert it to cartesian for Nav2, for example. This server is completely generic so that you can use it outside of Nav2 as well -- it really is fully generic for any F2C ROS 2 users and all their applications. I wrote extensive documentation of all the parameters and action options and its been tested at length in both GPS and cartesian environments. If you have confusion about how to handle GPS data, I suggest you pause on this server and take some time to learn that before continuing (certainly, I had to in my own education). 

Tl;dr:
- Its your responsibility to decide what frame you want to operate in and what the outcomes of that are for your system. Luckily its a one time setup that once you setup, will just work forever.
- The demo is just that: its a demo for a particular field, with a particularly known starting point, as a demonstration that the server works and how one could use it in the most simplistic case with Nav2. If you want to use GPS, another field, etc you need to adapt the demo as you would need to do for any demo.
- I provide the coverage path as well - not just the nav path - that contains the key points of interest. Many serious users of this work will be using the GPS coordinates of the row endings an then the turns rather than naively following a `nav_msgs/Path`.

**Comment by NK1807** [comment]
@SteveMacenski , Thanks for the detailed information.

> @NK1807 Did you find a solution for using the GPS coordinate with the plugin? Thanks in advice

@mattiadutto , Not really. But I found out that when the coordinates_in_cartesian_frame is set to false and given the gps coord gml file, the nav/msg/path published is in UTM frame(UTM coordinates) and the map frame is assigned for this data . So as @SteveMacenski said, a workaround need to be done. F2C need to be tweaked a bit to set frames that we need to work with. 
I didn't go deep into it. for time being, I'm converting lat/lon to map point using robot_localisation_node and sending map cartesian coordinates to the server.
If you would like to tweak F2C and make it work with GPS, I'm open to work together.

**Comment by SteveMacenski** [comment]
F2C works fine for GPS, you may not be providing the correct data in the right frame
